adapter/avro/src/main/java/org/apache/arrow/adapter/avro/AvroFileReader.java

```diff
@@ -0,0 +1,374 @@
+
+    // Initialize data vectors
+    List<FieldVector> vectors = new ArrayList<>(arrowSchema.getFields().size());
+    for (int i = 0; i < arrowSchema.getFields().size(); i++) {
+      FieldVector vector = recordConsumer.getConsumers().get(i).getVector();
+      vectors.add(vector);
+    }
+
+    // Initialize batch and schema
+    this.arrowBatch = new VectorSchemaRoot(vectors);
+    this.arrowSchema = arrowBatch.getSchema();
+
+    // First batch starts after the header
+    this.nextBatchPosition = headerSize;
+  }
+
+  private org.apache.avro.Schema processSchema(ByteBuffer buffer) throws IOException {
+
+    org.apache.avro.Schema.Parser parser = new org.apache.avro.Schema.Parser();
+
+    try (InputStream schemaStream = new ByteArrayInputStream(buffer.array())) {
+      return parser.parse(schemaStream);
+    }
+  }
+
+  private String processCodec(ByteBuffer buffer) {
+
+    if (buffer != null && buffer.remaining() > 0) {
+      return new String(buffer.array(), StandardCharsets.UTF_8);
+    } else {
+      return DataFileConstants.NULL_CODEC;
+    }
+  }
+
+  // Schema and VSR available after readHeader()
+  Schema getSchema() {
+    if (avroSchema == null) {
+      throw new IllegalStateException("Avro header has not been read yet");
+    }
+    return arrowSchema;
+  }
+
+  VectorSchemaRoot getVectorSchemaRoot() {
+    if (avroSchema == null) {
+      throw new IllegalStateException("Avro header has not been read yet");
+    }
+    return arrowBatch;
+  }
+
+  @Override
+  public Set<Long> getDictionaryIds() {
+    if (avroSchema == null) {
+      throw new IllegalStateException("Avro header has not been read yet");
+    }
+    return dictionaries.getDictionaryIds();
+  }
+
+  @Override
+  public Dictionary lookup(long id) {
+    if (avroSchema == null) {
+      throw new IllegalStateException("Avro header has not been read yet");
+    }
+    return dictionaries.lookup(id);
+  }
+
+  // Read the next Avro block and load it into the VSR
+  // Return true if successful, false if EOS
+  // Also false in non-blocking mode if need more data
+  boolean readBatch() throws IOException {
+
+    if (avroSchema == null) {
+      throw new IllegalStateException("Avro header has not been read yet");
+    }
+
+    if (!hasNextBatch()) {
+      return false;
+    }
+
+    // Read Avro block from the main encoder
+    long nRows = decoder.readLong();
+    batchBuffer = decoder.readBytes(batchBuffer);
+    decoder.readFixed(batchSyncMarker);
+
+    long batchSize =
+        zigzagSize(nRows)
+            + zigzagSize(batchBuffer.remaining())
+            + batchBuffer.remaining()
+            + SYNC_MARKER_SIZE;
+
+    // Validate sync marker - mismatch indicates a corrupt file
+    int validateMarker =
+        BinaryData.compareBytes(
+            syncMarker, 0, SYNC_MARKER_SIZE, batchSyncMarker, 0, SYNC_MARKER_SIZE);
+
+    if (validateMarker != 0) {
+      throw new RuntimeException("Invalid AVRO data file: The file is corrupted");
```

Comment:
ditto, can we throw a more specific exception? Maybe create a subclass of IOException to indicate invalid files?